src/rpc/conn_pool.rs

```diff
@@ -169,34 +168,32 @@ impl ConnPool {
             shared_pool: &Arc<SharedPool>,
             delay: Duration,
             min_build_retry_interval: Duration,
-            retry_num: usize,
+            mut retry_num: usize,
             build_retry_limit: usize,
         ) {
-            if retry_num > build_retry_limit {
-                let mut inner = shared_pool.inner.lock().unwrap();
-                inner.unpend_conn();
-                error!("ConnPool::add_connection_background::bg_add fail to build connection after {} retries", retry_num);
-                return;
-            }
-
-            let weak_shared_pool = Arc::downgrade(shared_pool);
-            shared_pool.conn_init_thread_pool.execute_after(delay, move || {
-                let shared_pool = match weak_shared_pool.upgrade() {
-                    None => return,
-                    Some(p) => p,
-                };
-
-                match shared_pool.build_conn() {
-                    Ok(conn) => {
-                        let mut inner = shared_pool.inner.lock().unwrap();
-                        inner.add_conn(conn);
-                        shared_pool.cond.notify_all();
-                    }
-                    Err(e) => {
-                        error!("ConnPool::add_connection_background::bg_add fail to build a connection after {} retries, err:{}", retry_num, e);
-                        let delay = cmp::max(min_build_retry_interval, delay);
-                        let delay = cmp::min(shared_pool.conn_builder.connect_timeout / 2, delay * 2);
-                        bg_add(&shared_pool, delay, min_build_retry_interval, retry_num + 1, build_retry_limit);
+            let shared_pool = shared_pool.clone();
```

Comment:
Why not remove the weak reference of `shared_pool`? If so, the shared_pool may not be dropped in some cases, e.g. where the conn_pool is dropped but the init procedure is sleeping at L195.

Reply:
Fixed